### DISCUSSION TITLE
Reducing noise by improving Issue grouping

### DISCUSSION THREAD

**Discussion by Dhrumil-Sentry** [getsentry/sentry#66319] [Upcoming Changes]
Ever so often Sentry notifies you that a ***new*** Issue was detected but you’ve already seen this Issue before - you see that this latest Issue has some minor differences in the stacktraces from the original one you viewed but it’s not a _**new**_ Issue

This problem can get even worse for various Javascript Issues - because there are parameterized URLs that show up in the stacktraces leading to numerous Issues getting created. We’re working on improving this experience for our users.

**We** **are actively seeking users who’d be happy to provide feedback on Issue Grouping as a part of our Beta cohort.** (Please reply below if you’re interested or drop us a note at feedback-issue-grouping@sentry.io

Here’s how we’re solving this problem in 2 phases:

## Detect Similar Issues more accurately

When you view an Issue in Sentry, you can go to the new ***Similar Issues*** tab and see a list of issues that are very similar based on stacktraces and should be grouped together based according to us:
<img width="668" alt="SimilarIssues" src="https://github.com/getsentry/sentry/assets/109983869/3ab126eb-e011-49aa-a7f0-0dbd8f26c633">

The “Would group” column represents Sentry’s opinion on whether these issue should have been grouped together instead. You can help determine the accuracy of our improvements by simply agreeing or disagreeing with these similar matchings
<img width="668" alt="SimilarIssues" src="https://github.com/getsentry/sentry/assets/109983869/8bcd4d2e-9188-4218-85a7-db1643a6a0a6">

This process allows us to identify cases where we’d over-group otherwise causing you to miss genuinely different issues and helps us QA the improvements

## Improving the grouping process itself

Once we validate that our enhancements can accurately determine issues that should be grouped together, we’ll start gradually shipping grouping improvements based on these enhancements to help reduce noise for our customers.

## We need your help

**We** **are actively seeking users who’d be happy to provide feedback on Issue Grouping as a part of our Beta cohort.** (Please reply below if you’re interested or drop us a note at feedback-issue-grouping@sentry.io

We’re actively looking for feedback on grouping problems you face while using Sentry and we’d appreciate inputs on any of the following:

1. What other grouping failures do you see occurring in your Sentry? (Would appreciate any links or images!)
2. What does your workflow look like when you find duplicate issues? Do you merge them? Bulk Archive or Resolve them?

**Comment by shaedrich** [comment]
Slightly unrelated but one of the problems, I have with the "similar issues" view is that it is a little cluttered. "Issues with similar stack trace" is definitely one grouping option, but there could be more, separate sections. These sections wouldn't need all the table rows but could focus on those that are important to the grouping option to see better at the first glance if they are indeed similar (enough).

**Reply by trillville** [reply]
That's related to one idea we've been thinking about lately, currently we're referring to it as "supergroups"

Curious, do you have any examples off the top of your head for what you'd be interested in seeing in those sections?

**Reply by shaedrich** [reply]
> All versions consider the fingerprint first, the stack trace next, then the exception, and then finally the message.
– _from the ["event/error grouping" docs](https://docs.sentry.io/product/data-management-settings/event-grouping/#default-error-grouping-algorithms)/["Grouping issues" docs](https://docs.sentry.io/product/issues/grouping-and-fingerprints/#changing-how-events-are-grouped)_

* **stack trace**: This is a tricky one, since stack traces can have a significant depth, so this might be framework-specific or related to [customized fingerprinting rules](https://docs.sentry.io/product/data-management-settings/event-grouping/fingerprint-rules/#matchers). But there are information, the stack trace has:
    * line
    * file
    * class _(optional)_
    * function/method _(optional)_
    * "block" (if condition, loop, etc.)
\
&rarr; Furthermore, there could be some more information, also provided in the [Trace navigator](https://docs.sentry.io/product/issues/issue-details/error-issues/#trace-navigator)
    * **Root**: Starting transaction
    * ~~**Ancestor**: Events that are descendants of the root, and are directly related to the current event~~
    * **Parent**: Event that directly started the current event
    * **This Event**: Node representing the current event being viewed
    * ~~**Children**: Events that the current event directly started~~
    * ~~**Descendants**: Future generations of events created by children of this event~~
* **exception**: This might have a similar problem to the stack trace in that it might depend on the parent class as well, if less 
* **message**: I always felt like `exception` and `message` are so tied together, they might be the same grouping option and `message` just being one level deeper within exception (essentially `exception.class` and `exception.message`). So the question here is more like whether `exception.class` AND `exception.message` have to be the same to be seen as similar or if one of them is enough
* **fingerprint**: Well, technically, this would need all the options below
    * [`error.type`](https://docs.sentry.io/product/data-management-settings/event-grouping/fingerprint-rules/#errortype)
    * [`error.value`](https://docs.sentry.io/product/data-management-settings/event-grouping/fingerprint-rules/#errorvalue)
    * [`message`](https://docs.sentry.io/product/data-management-settings/event-grouping/fingerprint-rules/#message)
    * [`logger`](https://docs.sentry.io/product/data-management-settings/event-grouping/fingerprint-rules/#logger)
    * [`level`](https://docs.sentry.io/product/data-management-settings/event-grouping/fingerprint-rules/#level)
    * [`tags.tag_name`](https://docs.sentry.io/product/data-management-settings/event-grouping/fingerprint-rules/#tagstag_name)
    * [`stack.abs_path`](https://docs.sentry.io/product/data-management-settings/event-grouping/fingerprint-rules/#stackabs_path)
    * [`stack.module`](https://docs.sentry.io/product/data-management-settings/event-grouping/fingerprint-rules/#stackmodule)
    * [`stack.function`](https://docs.sentry.io/product/data-management-settings/event-grouping/fingerprint-rules/#stackfunction)
    * [`stack.package`](https://docs.sentry.io/product/data-management-settings/event-grouping/fingerprint-rules/#stackpackage)
    * [`family`](https://docs.sentry.io/product/data-management-settings/event-grouping/fingerprint-rules/#family)
    * [`app`](https://docs.sentry.io/product/data-management-settings/event-grouping/fingerprint-rules/#app)

Hope, that helps

**Comment by deoomen** [comment]
Maybe an option to also manually merge issues will be good? For now we can merge issues only when Sentry mark them as similar. But sometimes Sentry does not see similarity but in out opinion there are the same issue.

For example:

```php
<?php

$array = null;

echo $array['b];
echo $array['c'];
```

Above example will produce two separate issues (`Warning: Trying to access array offset on null`) in Sentry but "Similar issues" tab is empty.

**Reply by Dhrumil-Sentry** [reply]
@deoomen Thanks for this feedback! Users can merge issues from the issue stream too- Does this not work?
![image](https://github.com/getsentry/sentry/assets/109983869/29030b01-8e9a-45cb-a8f1-a48e62694d36)

Would you consider the errors in your example to be the same issue even if they were occurring in slightly different stacktraces?

**Reply by deoomen** [reply]
Exactly. Stacktrace is slightly different but in this example it's only line number. And whole context is the same. So it'll be useful that we can merge those issues. Or maybe link somehow issues as related?

**Comment by spaceemotion** [comment]
I am indeed very much interested in better issue grouping. In my current project most of the errors are getting picked up incorrectly and thus make it very hard to filter out each error type on its own.

They all share the same stack trace, but the messages are different since they come from a 3rd party API.

**Reply by Dhrumil-Sentry** [reply]
@spaceemotion Thanks a lot for this feedback! 

> They all share the same stack trace, but the messages are different since they come from a 3rd party API.

Are these issues getting grouped into a single issue? Do you prefer them to be grouped as different issues?.

If you have any specific examples of such poorly grouped issues- please feel free to share the URLs to feedback-issue-grouping@sentry.io

**Reply by spaceemotion** [reply]
@Dhrumil-Sentry hey there, i wanted to send a mail, but i'm getting a `Delivery Status Notification (Failure)` from your mail service:

```
We're writing to let you know that the group you tried to contact (feedback-issue-grouping) may not exist, or you may not have permission to post messages to the group. A few more details on why you weren't able to post:

 * You might have spelled or formatted the group name incorrectly.
 * The owner of the group may have removed this group.
 * You may need to join the group before receiving permission to post.
 * This group may not be open to posting.

If you have questions related to this or any other Google Group, visit the Help Center at https://support.google.com/a/sentry.io/bin/topic.py?topic=25838.

Thanks,

sentry.io admins
```

**Reply by Dhrumil-Sentry** [reply]
@spaceemotion Sorry for the trouble ! I had to update a setting on the google group 🤦. Can you please resend 🙏

**Comment by rachrwang** [comment]
We're excited to announce that AI grouping has become generally available to all customers across all plans. This update leverages AI to group together semantically similar issues, and has dropped unnecessary issue creation by around 40%. 

To learn more about the engineering work behind this update, check out this post: https://blog.sentry.io/how-sentry-decreased-issue-noise-with-ai/
